8374307: Fix deoptimization storm caused by Action_none in GraphKit::uncommon_trap

bulasevich opened the pull request:
We observed a deoptimization storm caused by GraphKit::uncommon_trap generator logic. GraphKit::uncommon_trap considers the too_many_recompiles metric. If the threshold is overflowed, it replaces Deoptimization::Action_reinterpret with Deoptimization::Action_none (see code snippet below).

This replacement changes the uncommon_trap logic: once execution hits a trap, the VM performs deoptimization but does not recompile the method anymore. In an "unlucky" case, when the code part calling this uncommon_trap becomes frequent, a deoptimization storm occurs (thousands of deoptimizations per second) causing a significant performance drop.

The original problematic method, which triggered repeated recompilations, is a high-performance compressed binary serialization algorithm with heavy use of conditional branches driven by bitmasks. See a standalone [synthetic benchmark](https://bugs.openjdk.org/secure/attachment/118045/UnstableIf.java) to reproduce the issue.

The issue arises when the method overcomes a global recompilation threshold before stabilizing specific trap counters.

Current thresholds:
- Recompilation Limit (too_many_recompiles):
  Condition: decompile_count() >= (PerMethodRecompilationCutoff / 2) + 1
  Default: 201 (derived from default PerMethodRecompilationCutoff = 400).
- Specific Trap Limits (too_many_traps):
  Checks if the trap count for a specific reason exceeds:
  PerMethodTrapLimit (Default: 100) - for Reason_unstable_if, Reason_unstable_fused_if, etc.
  PerMethodSpecTrapLimit (Default: 5000) - for Reason_speculate_class_check, Reason_speculate_null_check, etc.

With the gived defaults, if the only reason for the method recompilation is unstable_if, the system stabilizes after 100 traps (PerMethodTrapLimit). However, if the method experiences traps and recompilations for different reasons, the total number of recompilations can exceed 200 before hitting the limit for unstable_if traps. This triggers Action_none and causes the deopt storm.

The proposal is a minimal change in GraphKit::uncommon_trap: apply the same `too_many_recompiles` threshold inside `Parse::path_is_suitable_for_uncommon_trap` - this ensures that on the final recompilation C2 gets a hint not to speculate on untaken branches anymore.

As an alternative solution, we can revisit GraphKit::uncommon_trap. This "Temporary fix" has persisted in the codebase for 17 years, so it is probably time to change it as well. Any comments are welcome
```cpp
  case Deoptimization::Action_reinterpret:
    // Temporary fix for 6529811 to allow virtual calls to be sure they
    // get the chance to go from mono->bi->mega
    if (!keep_exact_action &&
        Deoptimization::trap_request_index(trap_request) < 0 &&
        too_many_recompiles(reason)) {
      // This BCI is causing too many recompilations.
      if (C->log() != nullptr) {
        C->log()->elem("observe that='trap_action_change' reason='%s' from='%s' to='none'",
                Deoptimization::trap_reason_name(reason),
                Deoptimization::trap_action_name(action));
      }
      action = Deoptimization::Action_none;
      trap_request = Deoptimization::make_trap_request(reason, action);
    } else {
      C->set_trap_can_recompile(true);
    }
```

- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8374307](https://bugs.openjdk.org/browse/JDK-8374307): Fix deoptimization storm caused by Action_none in GraphKit::uncommon_trap (**Bug** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28966/head:pull/28966` \
`$ git checkout pull/28966`

Update a local copy of the PR: \
`$ git checkout pull/28966` \
`$ git pull https://git.openjdk.org/jdk.git pull/28966/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28966`

View PR using the GUI difftool: \
`$ git pr show -t 28966`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28966.diff">https://git.openjdk.org/jdk/pull/28966.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28966#issuecomment-3687555347)
</details>

Commit:
8374307: Fix deoptimization storm caused by Action_none in GraphKit::uncommon_trap

merykitty commented on src/hotspot/share/opto/parse2.cpp:
```diff
@@ -1618,6 +1617,8 @@ bool Parse::path_is_suitable_for_uncommon_trap(float prob) const {
   }
   return seems_never_taken(prob) &&
+         // Skip optimization if recompile limit is exceeded to avoid deopts without recompilation.
+         !C->too_many_recompiles(method(), bci(), Deoptimization::Reason_unstable_if) &&
```
You can use `Compile::too_many_traps_or_recompile` here.

TobiHartmann reviewed (commented):
Is this related to [JDK-8243615](https://bugs.openjdk.org/browse/JDK-8243615)? Could you convert your `UnstableIf.java` test to a jtreg test? Maybe by running in a different process and counting the number of deoptimization events? [JDK-8243615](https://bugs.openjdk.org/browse/JDK-8243615) also has a test attached.

Commit:
using too_many_traps_or_recompiles. adding DeoptStorm jtreg test

bulasevich commented:
> Is this related to [JDK-8243615](https://bugs.openjdk.org/browse/JDK-8243615)?

Oh, yes - this is related, and we had a similar fix five years ago.. @wzhuo

> Could you convert your `UnstableIf.java` test to a jtreg test? 

Done. I converted it to a jtreg test. I’m skipping the heavyweight part (200+ lines of code) that reproduces the issue without changing the PerMethodRecompilationCutoff limit.

iwanowww commented:
I believe all places where an uncommon trap with `Action_reinterpret` guarded by `too_many_traps` is susceptible to the very same problem.

The culprit seems to be the discrepancy between `too_many_traps` and `too_many_recompiles` where many places where uncommon traps are inserted are guarded by `too_many_traps` while `GraphKit::uncommon_trap()` checks specifically for `too_many_recompiles`.

As the bug demonstrates, disabling recompilation while keeping the uncommon trap in place (substituting `Action_maybe_recompile`/`Action_maybe_recompile` with `Action_none`) can induce a lot of overhead. So, a better strategy is to avoid an uncommon trap in the first place rather than letting it to degenerate into `Action_none` and, also, assert whenever the situation occurs at runtime.

Speaking of the proposed fix, my concern is that it addresses only one particular instance of the problem. Can we do better and fix similar bugs all at once? That would require aligning `too_many_traps` and `too_many_recompiles` use sites.

turbanoff commented on test/hotspot/jtreg/compiler/uncommontrap/DeoptStorm.java:
```diff
@@ -0,0 +1,104 @@
+            "-XX:CompileCommand=dontinline,compiler.uncommontrap::*",
+            "-Xlog:deoptimization=debug",
+            className, "dummy"};
+        ProcessBuilder pb  = ProcessTools.createLimitedTestJavaProcessBuilder(procArgs);
```
```suggestion
        ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(procArgs);
```

Commit:
Align uncommon-trap guards with recompilation cutoffs: replace too_many_traps() with too_many_traps_or_recompiles() at other susceptible sites

Commit:
minor cleanup

bulasevich commented:
@iwanowww, thank you!

> Can we do better and fix similar bugs all at once?

OK. With this change, I only use `too_many_traps_or_recompiles` to avoid further `uncommon_trap` in the too_many_recompiles case; this seems harmless and makes sense to backport to previous releases.

I think as followup it makes sense to eliminate deoptimization action replacement completely:
- https://github.com/openjdk/jdk/commit/729427f15f2bb2796323b9305339b94433ee74f6
- I will need help there with extensive testing, as jtreg and common benchmarks are not enough.

iwanowww commented:
Thanks, Boris!

> OK. With this change, I only use too_many_traps_or_recompiles to avoid further uncommon_trap in the too_many_recompiles case; this seems harmless and makes sense to backport to previous releases.

How did you choose which `too_many_traps` usages to change and which leave as is? I looked for `too_many_traps` usages and there are more than you touched.

merykitty commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -4514,7 +4527,7 @@ bool LibraryCallKit::inline_native_getLength() {
 // public static <T,U> T[] java.util.Arrays.copyOfRange(U[] original, int from,      int to, Class<? extends T[]> newType);
 bool LibraryCallKit::inline_array_copyOf(bool is_copyOfRange) {
-  if (too_many_traps(Deoptimization::Reason_intrinsic))  return false;
+  if (too_many_traps_or_recompiles(Deoptimization::Reason_intrinsic))  return false;
```
Since you are trying to replace all call sites of `too_many_traps` with `too_many_traps_or_recompiles` if the argument is `Action_reinterpret` / `Action_maybe_recompile`. I think it makes more sense to do it inside `GraphKit::uncommon_traps`. That is, to call `too_many_recompiles` if the argument is `Action_reinterpret` / `Action_maybe_recompile`. It also makes more sense to align `uncommon_trap` with `too_many_traps`, since that is the common pattern.

vnkozlov replied:
> I mean doing it inside GraphKit::too_many_traps

It may cause more changes since you will have to pass `deopt_action` to `too_many_traps()`. Also the name will not be correct because it will check recompilations.

We can add an other `GraphKit::too_many_traps_or_recompiles()` (or modify existing) and pass  `deopt_action` into it. Then check action and call `too_many_recompiles()` only for particular actions. And use it in all places where `too_many_traps()` is called.

bulasevich replied:
I understand the proposal as adding an action parameter:
```cpp
bool too_many_traps_or_recompiles(ciMethod* method, int bci, Deoptimization::DeoptReason reason,
								 Deoptimization::DeoptAction action) {
  if (too_many_traps(method, bci, reason)) {
    return true;
  }
  if (action == Deoptimization::Action_reinterpret ||
  	action == Deoptimization::Action_maybe_recompile) {
    return too_many_recompiles(method, bci, reason);
  }
  return false;
}
```

I'm not sure. The `too_many_traps_or_recompiles` check and the `uncommon_trap` call it guards are often far apart in the source, yet their action parameter need to remain consistent. Also, `too_many_traps_or_recompiles()` currently answers a question about the compilation history.
Once the intended action is taken into account, its meaning effectively becomes something like `too_many_traps_or_too_much_recompiles_and_still_going_to_recompile_again`, which looks like over-engineering.

bulasevich commented:
/touch

bulasevich commented:
/keepalive

schlosna commented:
Is there any support I could offer to help move this PR forward?

I recently ran across this deoptimization storm where `uncommon_trap` `isnonnull` triggers repeated deoptimization. I observed this by `jdk.Deoptimization` JFR events at [`com.fasterxml.jackson.databind.ser.BeanPropertyWriter` line 731 `if (_typeSerializer == null)` site](https://github.com/FasterXML/jackson-databind/blob/3c95742e2db8fbe360e7bedfa900e9d2b996c8d7/src/main/java/com/fasterxml/jackson/databind/ser/BeanPropertyWriter.java#L731-L735), and then re-enters `unstable_if` deopt thrash on the serialization path when a perturbing workload arrives leading to ~10x performance degradation and significant increases in CPU utilization.

bulasevich committed:
Fixing two more cases that can provole deopt storm:
1. Parse::create_jump_tables
-> too_many_traps(_or_recompiles)
-> jump_if_true_fork/jump_if_always_fork -> GraphKit::uncommon_trap: Action_reinterpret=>Action_none

2. PhaseStringOpts::PhaseStringOpts
-> build_candidate -> validate_control_flow -> too_many_traps_or_recompiles
-> replace_string_concat -> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

bulasevich commented:
Hi @iwanowww && @vnkozlov,

Just to recap the history of this PR: I initially fixed the code path where we reproduced the deoptimization storm. I then followed the review comments and updated other places where the same issue could occur: a too_many_traps check followed by a GraphKit::uncommon_trap call that may override the requested action:

GraphKit::new_array
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

GraphKit::add_parse_predicate
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

LibraryCallKit::inline_string_indexOfChar
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

LibraryCallKit::inline_Class_cast
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

LibraryCallKit::inline_native_getLength
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

LibraryCallKit::inline_array_copyOf
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

Parse::do_tableswitch
-> too_many_traps`_or_recompiles`
-> jump_switch_ranges -> jump_if_true_fork/jump_if_always_fork -> GraphKit::uncommon_trap: Action_reinterpret=>Action_none

Parse::do_lookupswitch
-> too_many_traps`_or_recompiles`
-> jump_switch_ranges -> jump_if_true_fork/jump_if_always_fork -> GraphKit::uncommon_trap: Action_reinterpret=>Action_none

Parse::create_jump_tables
-> too_many_traps`_or_recompiles`
-> jump_if_true_fork/jump_if_always_fork -> GraphKit::uncommon_trap: Action_reinterpret=>Action_none

Parse::adjust_map_after_if
-> path_is_suitable_for_uncommon_trap -> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_reinterpret=>Action_none

Parse::array_store_check
-> too_many_traps`_or_recompiles`
-> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

PhaseStringOpts::PhaseStringOpts
-> build_candidate -> validate_control_flow -> too_many_traps`_or_recompiles`
-> replace_string_concat -> GraphKit::uncommon_trap: Action_maybe_recompile=>Action_none

If you think this approach needs further refinement, I can take another pass at it.

bulasevich commented:
/touch
